File: java/triangle/src/main/java/Triangle.java
public class Triangle {
    private final TriangleKind kind;

    public Triangle(int x, int y, int z) throws TriangleException {
        this.kind = determineKind(x, y, z);
    }

    public Triangle(double x, double y, double z) throws TriangleException {
        this.kind = determineKind(x, y, z);
    }

    public TriangleKind getKind() {
        return this.kind;
    }

    private static TriangleKind determineKind(double x, double y, double z) throws TriangleException {
        if (x == 0 && y == 0 && z == 0)
        {
            throw new TriangleException();
        }

        if (x < 0 || y < 0 || z < 0)
        {
            throw new TriangleException();
        }

        if (x + y <= z || x + z <= y || y + z <= x)
        {
            throw new TriangleException();
        }

        if (x == y && y == z)
        {
            return TriangleKind.EQUILATERAL;
        }

        if (x == y || x == z || y == z)
        {
            return TriangleKind.ISOSCELES;
        }

        return TriangleKind.SCALENE;
    }
}

File: typescript/triangle/triangle.ts
export enum TriangleKind {
    Equilateral = "equilateral",
    Isosceles = "isosceles",
    Scalene = "scalene"
}

export default class Triangle {
    private x: number
    private y: number
    private z: number

    constructor(...sides: number[]) {
        [this.x, this.y, this.z] = sides.sort()
    }

    kind() {
        if (this.isInvalid()) {
            throw new Error('Invalid triangle')
        }

        if (this.isEquilateral()) {
            return TriangleKind.Equilateral
        }
        else if (this.isIsosceles()) {
            return TriangleKind.Isosceles
        }
        else {
            return TriangleKind.Scalene
        }
    }

    isInvalid() {
        return this.hasZeroSides() || this.hasNegativeSide() || this.violatesTriangleEquality()
    }

    private hasZeroSides() {
        return this.x === 0 && this.y === 0 && this.z === 0
    }

    private hasNegativeSide() {
        return this.x < 0 || this.y < 0 || this.z < 0
    }

    private violatesTriangleEquality() {
        return this.x + this.y <= this.z || this.x + this.z <= this.y || this.y + this.z <= this.x
    }

    private isEquilateral() {
        return this.x === this.y && this.y === this.z
    }

    private isIsosceles() {
        return this.x === this.y || this.y === this.z || this.x === this.z
    }
}